es/es-server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -319,4 +373,53 @@ static void buildShardLevelInfo(Logger logger, ShardStats[] stats, ImmutableOpen
             newShardSizes.put(sid, size);
         }
     }
+
+    static void fillDiskUsagePerNode(Logger logger, List<NodeStats> nodeStatsArray,
+            ImmutableOpenMap.Builder<String, DiskUsage> newLeastAvaiableUsages,
+            ImmutableOpenMap.Builder<String, DiskUsage> newMostAvaiableUsages) {
+        for (NodeStats nodeStats : nodeStatsArray) {
+            if (nodeStats.getFs() == null) {
+                logger.warn("Unable to retrieve node FS stats for {}", nodeStats.getNode().getName());
+            } else {
+                FsInfo.Path leastAvailablePath = null;
+                FsInfo.Path mostAvailablePath = null;
+                for (FsInfo.Path info : nodeStats.getFs()) {
+                    if (leastAvailablePath == null) {
+                        assert mostAvailablePath == null;
+                        mostAvailablePath = leastAvailablePath = info;
+                    } else if (leastAvailablePath.getAvailable().getBytes() > info.getAvailable().getBytes()){
+                        leastAvailablePath = info;
+                    } else if (mostAvailablePath.getAvailable().getBytes() < info.getAvailable().getBytes()) {
+                        mostAvailablePath = info;
+                    }
+                }
+                String nodeId = nodeStats.getNode().getId();
+                String nodeName = nodeStats.getNode().getName();
+                if (logger.isTraceEnabled()) {
```

Comment:
could we store this into a local var outside of the loop?

Reply:
I don't think so, since nodeStats is the object we are iterating on.